lib/puma.rb

```diff
@@ -10,14 +10,16 @@
 
 require 'thread'
 
+# use require, see https://github.com/puma/puma/pull/2381
 require 'puma/puma_http11'
-require 'puma/detect'
-require 'puma/json_serialization'
+
+require_relative 'puma/detect'
+require_relative 'puma/json_serialization'
 
 module Puma
-  autoload :Const, 'puma/const'
-  autoload :Server, 'puma/server'
-  autoload :Launcher, 'puma/launcher'
+  autoload :Const,    "#{__dir__}/puma/const"
```

Comment:
This feels weird. Is this weird? I've never seen this before.

Reply:
> This feels weird. Is this weird? I've never seen this before.

If it's an absolute path, it 'bypasses' the RubyGems resolution.  So

`autoload :Const, "puma/const"` resolved by RubyGems, similar to `require`
`autoload :Const, "#{__dir__}/puma/const"` resolved as a path, no RubyGems, similar to `require_relative`

The Ruby docs for `Kernel.autoload` show an absolute path, `Module.autoload` show a 'RubyGems' like path.

I think most code paths that are followed with 'normal' Puma use will see `require_relative` for all the files autoloaded here, so these are kind of helpers for running single test files.  `autload` only does something if the constant isn't defined...